jupyter/src/askee_kernel/askeekernel.py

```diff
@@ -7,61 +7,102 @@
 from typing import Dict
 import traceback
 import tabulate
-import requests
 from ipykernel.kernelbase import Kernel
+import json
+import websocket
+import threading
+import queue
 
 #------------------------------------------------------------------------------
 # Config
 #------------------------------------------------------------------------------
-DONU_ADDRESS = "http://localhost:8000"
+DONU_WS_ADDRESS = "ws://localhost:8000/:exposure"
 vega_lite_schema = "https://vega.github.io/schema/vega-lite/v4.json"
 
+
 # -----------------------------------------------------------------------------
 # Donu client
 # -----------------------------------------------------------------------------
 
+def on_message(resp_q):
+    """
+    This just adds any websocket messages to our
+    internal response queue so that they can be
+    processed by the protocol thread
+    """
+    def go(app, msg):
+        resp_q.put(msg)
+    return go
+
+def on_open(req_q, resp_q, kernel_q):
+    """
+    Starts the protocol thread on connect
+    """
+    def go(app):
+        threading.Thread(target=exposure_protocol, args=[app, req_q, resp_q, kernel_q]).start()
+    return go
+
+def exposure_protocol(app, req_q, resp_q, to_kernel_q):
+    """
+    Implements the exposure protocol for the client.
+    This looks like:
+    1. Client sends code to donu for evaluation
+    2. On response,
+      a. the response is a request for a file: we read the file and send it back, go to 2
+      b. the response is a success or failure: enqueue the response on the kernel queue.
+    """
+    while True:
+        app.send(req_q.get())
+        while True:
+            resp = json.loads(resp_q.get())
+
+            if resp['type'] == 'read-file':
+                with open(resp['path']) as f:
+                    app.send(json.dumps({'type':'file-contents', 'contents': f.read()}))
+
+            elif resp['type'] == 'write-file':
+                with open(resp['path'], 'w+') as f:
+                    f.write(resp['contents'])
+
+            elif resp['type'] == 'success':
+                break
```

Comment:
Both `'success'` and `'failure'` cause a `break` to happen. What is the difference between these responses?

Reply:
For the purposes of the protocol with the server, there is no difference. They are interpreted later, by `execute_donu_command` at line 336.

The idea is:
(1) client (kernel) sends an exposure program to the server that may reference the filesystem.
(2) the server will ask the client for any files it needs while it interprets the program from (1). When this is done, the server will send either 'success' or 'failure', indicating that the evaluator has terminated.
(3) the result of (2) is passed to the caller of `execute_code` via the `kernel_q` (which happens to be line `336`).

I tried to capture the protocol in the docstring above, but if this explanation helps then I'll add it.

Reply:
Ah, I missed the fact that `'success'` and `'failure'` trigger behavior elsewhere. In that case, I'm happy with this, and I like the description in the docstring.